torch/distributed/fsdp/_optim_utils.py

```diff
@@ -1329,3 +1377,100 @@ def return_fn(fqn_to_param_info):
         return_fn,
         fqn_to_param_info,
     )
+
+
+def _gather_orig_param_state(
+    fqn: str,
+    fsdp_param_info: FSDPParamInfo,
+    optim_state: Dict[str, Any],
+    shard_state: bool,
+) -> Dict[str, Any]:
+    """
+    Gather the optimizer state for the original parameter with the name ``fqn``.
+    This API should only be used when ``use_orig_params`` is True.
+    """
+    fsdp_state = fsdp_param_info.state
+    flat_param = fsdp_param_info.flat_param
+    param_idx = fsdp_param_info.param_indices[fqn]
+    if (
+        fsdp_state.world_size == 1
+        or fsdp_state.sharding_strategy == ShardingStrategy.NO_SHARD
+    ):
+        return optim_state
+
+    # Gathering state from all ranks. This step may be slow. However,
+    # `state_dict()` is not in the critical path. We can fuse the communication
+    # if the performance becomes a problem.
+    state_objects = {
+        state_name: value for state_name, value in sorted_items(optim_state)
+    }
+    object_list: List[Dict[str, Any]] = [
+        {} for _ in range(cast(int, fsdp_state.world_size))
+    ]
+    dist.all_gather_object(object_list, state_objects)
+    orig_state: Dict[str, Any] = {}
+    for state in object_list:
+        for state_name, value in state.items():
+            curr_value = orig_state.get(state_name, [])
+            if torch.is_tensor(value):
+                if value.dim() > 0:
+                    curr_value.append(value)
+                elif torch.is_tensor(curr_value):
+                    assert torch.equal(curr_value, value)
+                else:
+                    orig_state[state_name] = value
+            else:
+                assert curr_value == [] or curr_value == value
+                orig_state[state_name] = value
+
+    # Unflatten state values.
+    for state_name in orig_state.keys():
+        value = orig_state[state_name]
+        if not isinstance(value, list) or not torch.is_tensor(value[0]):
+            continue
+        value = torch.concat(value)[: flat_param._numels[param_idx]].reshape(
+            flat_param._shapes[param_idx]
+        )
+        if shard_state:
+            value = _ext_chunk_tensor(
+                value,
+                cast(int, fsdp_state.rank),
+                cast(int, fsdp_state.world_size),
+                torch.cuda.device_count(),
+                cast(dist.ProcessGroup, fsdp_state.process_group),
+            )
+        value = value.cpu()
+        orig_state[state_name] = value
+    return orig_state
+
+
+def _shard_orig_param_state(
+    fqn: str,
```

Comment:
nit: For this function and `_gather_orig_param_state`, can we specify if `fqn` is the local FQN (as stored in `flat_param._fqns`) or the global FQN (which may require prepending a prefix starting from the local FSDP root).

Ugh, we might need to clarify our terminology at some point and just write it down somewhere.